lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,496 @@
+				if len(separatedBatches[eventType]) != 0 {
+					watcher.notifyModelChan <- separatedBatches[eventType]
+				}
+			}
+			// If sending to channel blocked for a long time,
+			// shorten next notifyDelay accordingly.
+			duration := time.Since(timeBeforeSending)
+			buffer := time.Duration(1) * time.Millisecond
+			switch {
+			case duration < watcher.notifyDelay/10:
+				watcher.resetNotifyTimerChan <- watcher.notifyDelay
+			case duration+buffer > watcher.notifyDelay:
+				watcher.resetNotifyTimerChan <- buffer
+			default:
+				watcher.resetNotifyTimerChan <- watcher.notifyDelay - duration
+			}
+		}()
+		return
+	}
+	l.Debugln(watcher, "No old fs events")
+	watcher.resetNotifyTimer(watcher.notifyDelay)
+}
+
+func (watcher *fsWatcher) popOldEvents(dir *eventDir, currTime time.Time) fsEventsBatch {
+	oldEvents := make(fsEventsBatch)
+	for _, childDir := range dir.dirs {
+		for path, event := range watcher.popOldEvents(childDir, currTime) {
+			oldEvents[path] = event
+		}
+	}
+	for path, event := range dir.events {
+		// 2 * results in mean event age of notifyDelay (assuming randomly
+		// occurring events). Reoccuring and remove (for efficient renames)
+		// events are delayed until notifyTimeout.
+		if (event.eventType != remove && 2*currTime.Sub(event.lastModTime) > watcher.notifyDelay) ||
+			currTime.Sub(event.firstModTime) > watcher.notifyTimeout {
+			oldEvents[path] = event
+			delete(dir.events, path)
+		}
+	}
+	if dir.parentDir != nil && dir.childCount() == 0 {
+		dir.parentDir.removeEmptyDir(dir.path)
+	}
+	return oldEvents
+}
+
+func (watcher *fsWatcher) updateInProgressSet(event events.Event) {
+	if event.Type == events.ItemStarted {
+		path := event.Data.(map[string]string)["item"]
+		watcher.inProgress[path] = struct{}{}
+	} else if event.Type == events.ItemFinished {
+		path := event.Data.(map[string]interface{})["item"].(string)
+		delete(watcher.inProgress, path)
+	}
+}
+
+func (watcher *fsWatcher) pathInProgress(path string) bool {
+	_, exists := watcher.inProgress[path]
+	return exists
+}
+
+func (watcher *fsWatcher) UpdateIgnores(ignores *ignore.Matcher) {
+	l.Debugln(watcher, "Ignore patterns update")
+	watcher.ignoresUpdate <- ignores
+}
+
+func (watcher *fsWatcher) String() string {
+	return fmt.Sprintf("fswatcher/%s:", watcher.description)
+}
+
+func (dir *eventDir) eventCount() int {
+	count := len(dir.events)
+	for _, dir := range dir.dirs {
+		count += dir.eventCount()
+	}
+	return count
+}
+
+func (dir *eventDir) childCount() int {
+	return len(dir.events) + len(dir.dirs)
+}
+
+func (dir *eventDir) removeEmptyDir(path string) {
+	delete(dir.dirs, path)
+	if dir.parentDir != nil && dir.childCount() == 0 {
+		dir.parentDir.removeEmptyDir(dir.path)
+	}
+}
+
+func watchesLimitTooLowError(folder string) error {
+	return errors.New("Failed to install inotify handler for " +
+		folder +
+		". Please increase inotify limits," +
+		" see http://bit.ly/1PxkdUC for more information.")
```

Comment:
This probably should point to an FAQ link or our docs link, not an address in Libya.